server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1589,6 +1509,148 @@ static String resolveExpression(String expression, LongSupplier getTime) {
+                validateAliasOrIndex(expression);
+                if (expression.isWildcard() || expression.isExclusion() || ensureAliasOrIndexExists(context, expression.toString())) {
+                    result.add(expression.isExclusion() ? "-" + expression : expression.toString());
+                }
+            }
+            return result;
+        }
+
+        @Nullable
+        private static boolean ensureAliasOrIndexExists(Context context, String expression) {
+            boolean ignoreUnavailable = context.getOptions().ignoreUnavailable();
+            IndexAbstraction indexAbstraction = context.getState().getMetadata().getIndicesLookup().get(expression);
+            if (indexAbstraction == null) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    throw notFoundException(expression);
+                }
+            }
+            // treat aliases as unavailable indices when ignoreAliases is set to true (e.g. delete index and update aliases api)
+            if (indexAbstraction.getType() == Type.ALIAS && context.getOptions().ignoreAliases()) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    throw aliasesNotSupportedException(expression);
+                }
+            }
+            if (indexAbstraction.isDataStreamRelated() && context.includeDataStreams() == false) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    IndexNotFoundException infe = notFoundException(expression);
+                    // Allows callers to handle IndexNotFoundException differently based on whether data streams were excluded.
+                    infe.addMetadata(EXCLUDED_DATA_STREAMS_KEY, "true");
+                    throw infe;
+                }
+            }
+            return true;
+        }
+
+        private static void validateAliasOrIndex(ExpressionIterable.Expression expression) {
+            if (expression.isExclusion()) {
+                return;
+            }
+            if (Strings.isEmpty(expression.toString())) {
+                throw notFoundException(expression.toString());
+            }
+            // Expressions can not start with an underscore. This is reserved for APIs. If the check gets here, the API
+            // does not exist and the path is interpreted as an expression. If the expression begins with an underscore,
+            // throw a specific error that is different from the [[IndexNotFoundException]], which is typically thrown
+            // if the expression can't be found.
+            if (expression.toString().charAt(0) == '_') {
+                throw new InvalidIndexNameException(expression.toString(), "must not start with '_'.");
+            }
+        }
+    }
+
+    public static final class ExpressionIterable implements Iterable<ExpressionIterable.Expression> {
```

Comment:
This encapsulates the logic that "expressions starting with - following another expression starting with * are exclusions".

Reply:
I think this is a bit hard to read mostly due to the parallel nature of idx and i in list.get(i). Reading between the lines...are you trying to avoid new String allocations by adding parallel information to a complementary data structure ? 

...if so, I am not sure that is an optimization is worth the effort over just creating new Expressions that wrap the String and other other requisite meta data.  Also, I don't think the laziness of the iterable buys us much where an alternative would be proactively create a different data structure i..e. a list of Expressions that have flags for is_wildcard, is_subtraction, etc. explicitly created all at once from the list of Strings. It might be a tiny bit less performant, but would be a much easier model and to reason about.  

for example (not tested ...just an an example), 

```
AtomicBoolean hasWildCard = new AtomicBoolean(false);
List<Expression> expressions = original.stream()
        .map(o -> new Expression(o, hasWildCard.compareAndSet(!isWildCard(o), true), o.startsWith("-"))).toList();
new Expressions(hasWildCard, expressions);      


record Expression(String expression, boolean isWildcard, boolean isSubtraction) {}

record Expressions(AtomicBoolean hasWildCard, List<Expression> expressions){}
```

Reply:
Sort of yes, I am trying to not create another list that's only used for a single iteration.
I see your suggestion is to build the `List<Expression>`. I can implement that.

Reply:
> eading between the lines...are you trying to avoid new String allocations by adding parallel information to a complementary data structure ?

You infered correctly. I was trying to not build another list of records which contain another String for the expression.
If you find this hard to read, I agree it's probably unnecesary optimisation.
I have pushed https://github.com/elastic/elasticsearch/pull/92518/commits/2f4aa15b05e7dd47f9eabea4c5246aa48322bab5 that eagerly builds the expression list (where elements have the`isWildcard` and `isExclusion` properties).